### PR TITLE
cache: Redis integration

### PR DESCRIPTION
**Pull request by pierre**
All tests using H2 and Redis (for both the clock and caching) pass.

### CODE REVIEW

File: util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.Set;
+
+import javax.cache.Cache.Entry;
+
+import org.apache.shiro.cache.Cache;
+import org.apache.shiro.cache.CacheException;
+import org.apache.shiro.cache.CacheManager;
```
Comment (sbrossie) [inline util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java]: I am not sure i follow exactly what is happening. I would have expected to import `javax.cache.CacheManager` not `org.apache.shiro.cache.CacheManager`, but i am guessing `RedisShiroManager` manages the Shiro cache using redis as an implementation?

File: util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        @Override
+        public V get(final K key) throws CacheException {
+            return cache.get(key);
+        }
+
+        @Override
+        public V put(final K key, final V value) throws CacheException {
+            V previousValue;
+            while (true) {
+                previousValue = cache.get(key);
```
Comment (sbrossie) [inline util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java]: Confused too

File: util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java (hunk 1 above)
Comment (pierre) [reply util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java]: The `RedisShiroManager` indeed implements a `Shiro` `org.apache.shiro.cache.CacheManager`. `Shiro` doesn't support `javax.cache.CacheManager` and has its own thing instead, so this does the bridge between the two cache managers.

The `javax.cache.CacheManager` implementation for Redis is done via `Redis107CacheManagerProvider` (it's a `org.redisson.jcache.JCacheManager`).

File: util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java (hunk 2 above)
Comment (pierre) [reply util/src/main/java/org/killbill/billing/util/glue/RedisShiroManager.java]: Which part specifically?

**Comment by pierre** [comment]
Merging to make subsequent PR easier to read, but happy to continue the conversation above.

**Pull request merged**